### PR TITLE
Fix DoF's dependency on resolution

### LINKED ISSUE #56911: Vulkan: CameraEffects depth of field amount is resolution-dependent
### Godot version

4.0.dev (ce6de717f)

### System information

Fedora 34, GeForce GTX 1080 (NVIDIA 495.46)

### Issue description

*Related to https://github.com/godotengine/godot/issues/54518.*

The DoF effect's amount is resolution-dependent. This means the depth of field will appear to be *stronger* as the rendering resolution decreases.

This means the rendering will change in unintended ways depending on the user's monitor resolution. Also, if we implement dynamic resolution scaling and [FSR](https://github.com/godotengine/godot-proposals/issues/2809) in the future, we need to make sure every effect is as a resolution-independent as possible.

https://user-images.githubusercontent.com/180032/149983514-1ea30fa8-9640-4139-b143-19e56a2597ce.mp4

### Steps to reproduce

- Add a Camera3D node with a CameraEffects resource.
- Enable Dof Far and Dof Near on the CameraEffects resource.
- Create a MeshInstance with a SphereMesh. Move the SphereMesh away from the camera so the depth of field can be visible from the Camera3D preview.
- Add a StandardMaterial3D on the SphereMesh, with a bright albedo and/or emission value (to make depth of field visible).
- Decrease the 3D resolution scale factor at run-time using `get_viewport().scale_3d -= 0.1`.

### Minimal reproduction project

https://github.com/Calinou/test_fxaa_supersample_master
Press the <kbd>Up</kbd> / <kbd>Down</kbd> arrow keys to adjust the 3D rendering scale. When doing so, watch the depth of field become stronger as the resolution decreases.

### PR DESCRIPTION
**Pull request by ghost**
fixes #56911

This has no impact on how the glow effect scales with the resolution, but I couldn't be bothered to do a different recording.

https://user-images.githubusercontent.com/9743949/151022884-b3a9d490-066e-44aa-a377-ca5f74a59236.mp4

### CODE REVIEW

**Comment by clayjohn** [comment]
This looks fine to me, but I am worried about the performance impact, especially when using the circular bokeh mode. 

Can you test circular bokeh on both low res, HD, and 4K?

Or maybe @Calinou is able to test on HD and 4K?

File: servers/rendering/renderer_rd/renderer_scene_render_rd.cpp [hunk 1]
```diff
@@ -2362,6 +2362,8 @@ void RendererSceneRenderRD::_render_buffers_post_process_and_tonemap(const Rende
 
 	// @TODO IMPLEMENT MULTIVIEW, all effects need to support stereo buffers or effects are only applied to the left eye
 
+	float resolution_factor = p_render_data->cam_vaspect ? (float(rb->internal_width) / 1920.0) : (float(rb->internal_height) / 1080.0);
```
Comment (akien-mga) [inline servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:2365]: Does 1920x1080 correspond to something specific in the renderer? It seems a bit weird to see this hardcoded, wouldn't it need to reflect the actual game viewport size?

Comment (akien-mga) [reply servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:2365]: Ah I see this was discussed in https://github.com/godotengine/godot/issues/56911#issuecomment-1019330478. A comment to justify the hardcoded reference size would be useful.

**Comment by ceLoFaN** [comment]
I tested with Circular Bokeh and High Quality.

With this patch:
| Scaling                              | FPS       |
| ---------------------------------- | ----------- |
| 0.5 scaling (960x540)       | 480 |
| 1.0 scaling (1920x1080)   | 195 |
| 2.0 scaling (3840x2160):  | 18   |

Without this patch:
| Scaling                              | FPS       |
| ---------------------------------- | ----------- |
| 0.5 scaling (960x540)       | 395 |
| 1.0 scaling (1920x1080)   | 195 |
| 2.0 scaling (3840x2160):  | 55   |

Picking a higher reference resolution would most likely improve the situation for 4k as well.

File: servers/rendering/renderer_rd/renderer_scene_render_rd.cpp (hunk 1 above)
Comment (ceLoFaN) [reply servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:2365]: I agree, we probably shouldn't hardcode such values. @Calinou touched upon this point in [his comment](https://github.com/godotengine/godot/issues/56911#issuecomment-1019330478).

**Comment by clayjohn** [comment]
> Picking a higher reference resolution would most likely improve the situation for 4k as well, but I'm not sure that's the solution or if one is required. Any suggestions?

Picking a higher reference resolution is the same as decreasing the fixed ``64.0`` multiplier or reducing the default bokeh size (which was likely chosen for artistic reasons rather than performance reasons).

I see two solutions and I am torn between the two:
1. Go with this PR and have a warning in the documentation that performance will scale with resolution, so at high resolutions, you may want to reduce bokeh size proportionately.
2. Don't merge this PR and have a warning saying that Bokeh size is fixed at 64 pixels * user_specified_size and users should scale this with resolution if they want consistent results. 

Number 1 is nice, because we want the default to be "just works". That being said the performance cost is pretty hard to swallow. The end result would be "it just works*" (* but runs too slow to use).

Number 2 is better for performance and for users who don't care about blur amount as much. 

Can you do another table like the above, but for Box blur and hexagon blur? They both should scale much better with resolution. If the difference isn't as bad with them, then I lean towards 1.

**Comment by Calinou** [comment]
> Can you do another table like the above, but for Box blur and hexagon blur? They both should scale much better with resolution. If the difference isn't as bad with them, then I lean towards 1.

I feel we should change the circle bokeh shape to *not* have performance that depends on resolution, even if this means it'll look worse at higher resolutions. The caveat is just too bothersome for developers to handle, especially if they want to expose circle bokeh as an opt-in setting for those with fast GPUs.

**Comment by ceLoFaN** [comment]
I wrote a little Python function to understand how many iterations are being made for the circular bokeh inside [this for loop](https://github.com/godotengine/godot/blob/583b6a594ab88208d1e78c94e70a2957c316cf64/servers/rendering/renderer_rd/shaders/bokeh_dof_raster.glsl#L204):
```python
def a(blur_scale, blur_size):
  radius = blur_scale
  c = 0
  while radius < blur_size:
    radius += blur_scale / radius
    # print(radius)
    c += 1
  return c
```

`blur_scale` = one of `[8.0, 4.0, 1.0, 0.5]` - it represents the [blur quality setting](https://github.com/godotengine/godot/blob/58324f4df822c7416278cda5bf41662e979f9e31/servers/rendering/renderer_rd/effects_rd.cpp#L1141), from `very low quality` to `high quality`

`blur_size` = `dof_blur_amount (from the camera)` * `64.0 (hardcoded multiplier)` * `resolution_factor (from this PR)`

----

And here's some results:

| blur_scale | blur_size | iterations |
| ------------ | ----------- | --------------|
| 8 (VLQ)      | 0.2 * 64 * 1.0 | 6 |
| 4 (LQ)      | 0.2 * 64 * 1.0 | 18 |
| 1 (MQ)      | 0.2 * 64 * 1.0 | 80 |
| 0.5 (HQ)      | 0.2 * 64 * 1.0 | 162 |
| - | - | - |
| 8 (VLQ)      | 0.2 * 64 * 2.0 | 37 |
| 4 (LQ)      | 0.2 * 64 * 2.0 | 79 |
| 1 (MQ)      | 0.2 * 64 * 2.0 | 326 |
| 0.5 (HQ)      | 0.2 * 64 * 2.0 | 653 |
| - | - | - |
| 0.5 (HQ)      | 0.1 * 64 * 1.0 | 39 |
| 0.5 (HQ)      | 0.2 * 64 * 1.0 | 162 |
| 0.5 (HQ)      | 0.3 * 64 * 1.0 | 367 |
| 0.5 (HQ)      | 0.4 * 64 * 1.0 | 653 |
| 0.5 (HQ)      | 0.5 * 64 * 1.0 | 1022 |

The increase in iterations looks pretty steep when cycling between DoF amount and qualities.
As far as I can tell these numbers (also) don't depend on the resolution, and maybe they should?

**Comment by clayjohn** [comment]
> The increase in iterations looks pretty steep when cycling between DoF amount and qualities.
> As far as I can tell these numbers (also) don't depend on the viewport's resolution, and maybe they should?

Your resolution multiplier would make this dependent on resolution. You can see the default setting is to blur over 64 pixels * dof_amount. 

If we wanted to keep the cost the same as screen size changed we could scale both the ``blur_scale`` and ``blur_size`` This would keep the size and performance constant, but quality would suffer with a change in resolution.

**Comment by ceLoFaN** [comment]
> Your resolution multiplier would make this dependent on resolution. You can see the default setting is to blur over 64 pixels * dof_amount.

Yes, I was thinking about `blur_scale`, but wasn't obvious from what I wrote.

> If we wanted to keep the cost the same as screen size changed we could scale both the blur_scale and blur_size This would keep the size and performance constant, but quality would suffer with a change in resolution.

I'll try to find a sweet spot where quality doesn't visibly change much.

**Comment by ceLoFaN** [comment]
As promised, I've been working on improving the performance of the circle blur. Here's some screenshots to compare performance and looks:
| master | optimised |
| --------- | ------------- |
| dof_master_circle_hq_09 | dof_branchv6_circle_hq_1 |

I've still got a few things to sort out, like improving the quality and performance of the near blur, maybe see if I can make the whole thing resemble the unoptimised effect better... and tidy up the code, although it's mostly just some changes inside the shader code.

I'm thinking of introducing these tweaks as a new DoF mode and call it "Circle Optimised" (?). So whoever wants to use the high quality, uncompromising classic "Circle" mode can still do so, but I think that one is more suitable for offline rendering as it stands.

---

Even if the optimized blur performs well regardless of `blur_size`, scaling up the resolution increases both the pixel count and the `blur_size` (as of this PR)... so performance will degrade quicker when scaling, no matter how optimized the effect is.

That means I might still have to add a setting, for a finer control of the decrease in quality as resolution goes up.

I could also just pick a default resolution-quality factor that's good enough, and call it a day.

**Comment by Calinou** [comment]
> So whoever wants to use the high quality, uncompromising classic "Circle" mode can still do so, but I think that one is more suitable for offline rendering as it stands.

I think your new mode can replace the existing Circle mode, as offline rendering is considered a corner case. It probably doesn't warrant having entire rendering modes on its own.

**Comment by Calinou** [comment]
@ceLoFaN Could you push your changes with the optimized circle DoF?

**Comment by ceLoFaN** [comment]
Sorry for the long silence, but I've been dealing with various health issues for the past 5 weeks or so.

Here's the branch if anybody wants to tidy it up or give it a spin: https://github.com/godotengine/godot/compare/master...ceLoFaN:improve-dof?expand=1

Some numbers might need more tweaking to get the before advertised FPS, but the good news is I fixed 2 bugs that resulted in visual artifacts at various DOF settings, and it resembles the original DOF better.

I was also messing with a denoising step at the time, which you guys might not like as much, but to me looked somewhat promising.

Unfortunately, I don't think I'll be able to resume work on this for the forseable future, but I pushed what I had at the time on my working branch.

**Pull request closed without merging**